library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/types/ImmutableAbstractMessage.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2023 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.chat.api.types;
+
+import net.minecraft.entity.player.PlayerEntity;
+import org.jetbrains.annotations.NotNull;
+import org.quiltmc.qsl.chat.api.QuiltMessageType;
+
+import java.util.EnumSet;
+
+/**
+ * @param <T> The type of the immutable type
+ * @param <S> The type of the packet form of this class
+ */
+sealed public abstract class ImmutableAbstractMessage<T extends ImmutableAbstractMessage<T, S>, S>
+		permits ImmutableC2SChatMessage, ImmutableS2CChatMessage, ImmutableS2CSystemMessage, MutableAbstractMessage {
+	protected final @NotNull PlayerEntity player;
+	protected final boolean isOnClientSide;
+
+	protected ImmutableAbstractMessage(@NotNull PlayerEntity player, boolean isOnClientSide) {
+		this.player = player;
+		this.isOnClientSide = isOnClientSide;
+	}
+
+	// These are mostly annotated as @NotNull for dev productivity, I get nice warnings about
+	// What parts of the subclasses I need to implement still
+	abstract public @NotNull EnumSet<QuiltMessageType> getTypes();
+
+	abstract public @NotNull T immutableCopy();
+
+	abstract public @NotNull S asPacket();
```

Comment:
```suggestion
	public abstract @NotNull EnumSet<QuiltMessageType> getTypes();

	public abstract @NotNull T immutableCopy();

	public abstract @NotNull S asPacket();
```
excuse me ma'am, but it appears your modifiers are on backwards